File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        def binarysearch(low,high):
            
            while low<=high:          
                mid = (low + high)//2
                if nums[mid] == target: return mid
                elif target < nums[mid] : high = mid -1
                else: low = mid +1
                
            return -1
                
        def minelement():
            
            low =0
            size = len(nums)
            high = size -1
            
            if size ==1: return nums[0]
            
            while low <= high:

                mid = (low+high)//2
                n = (mid +1)%size
                p = (mid -1 + size) %size
                
                
                if nums[mid] < nums[n] and nums[mid] < nums[p]: return mid
                elif nums[mid] < nums[high]: high = mid -1
                else: low = mid +1
                
            return -1
        
        mini = minelement()
        print(mini)
        a = binarysearch(0,mini-1)
        b = binarysearch(mini,len(nums)-1)
        
        if a!=-1 : return a
        else : return b
            
        

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    
    int binarysearch(vector<int>&nums, int target,int s ,int e){
        
        while(s<=e){
            
            int mid = s + (e-s)/2;
            
            if(nums[mid]< target) s = mid+1;
            else if( nums[mid] > target) e = mid-1;
            else return mid;
        }
        
        return -1;
    }
    
    int findmin(vector<int>&nums){
        
        int n = nums.size();
        int s= 0;
        int e = n-1;
        
        if(n == 1) return 0;
        
        while(s<=e){
            
            int mid  = s + (e-s)/2;
            
            int prev = (mid -1+n)%n;
            int next = (mid +1)%n;
            
            if(nums[mid] < nums[prev] and nums[mid] < nums[next]) return mid;
            if(nums[mid] < nums[n-1]) e = mid-1;
            else s = mid+1;
        }
        
        return -1;
    }
    int search(vector<int>& nums, int target) {
        
        int min = findmin(nums);
        int left = binarysearch(nums,target, 0,min-1);
        int right = binarysearch(nums,target,min,nums.size()-1);
        
        if(left != -1) return left;
        else if(right != -1) return right;
        else return -1;
        
        return 0;
        
    }
};